rfcs/20200710-reduction/README.md

```diff
@@ -0,0 +1,236 @@
+src = binary(div, src, l2_norm)
+```
+
+## Option 1: reduction and norms as algorithms of single primitive (Recommended)
+
+### API
+``` cpp
+// dnnl_types.h
+typedef enum {
+    ...
+    dnnl_reduction_max,
+    dnnl_reduction_min,
+    dnnl_reduction_norm_lp_max,
+    dnnl_reduction_norm_lp_add,
+    dnnl_reduction_norm_lp_power_p_max,
+    dnnl_reduction_norm_lp_power_p_add,
+    dnnl_reduction_sum,
+    dnnl_reduction_mul,
+    dnnl_reduction_mean,
+    ...
+} dnnl_alg_kind_t;
+
+typedef struct {
+    dnnl_primitive_kind_t primitive_kind;
+    /// The kind of reduction algorithm.
+    dnnl_alg_kind_t reduction_alg_kind;
+    /// Source memory descriptor.
+    dnnl_memory_desc_t src_desc;
+    /// Destination memory descriptor.
+    dnnl_memory_desc_t dst_desc;
+    /// Algorithm specific parameters.
+    /// Accordance table:
+    /// #dnnl_reduction_max: @p p and @eps are ignored
+    /// #dnnl_reduction_min: @p p and @eps are ignored
+    /// #dnnl_reduction_norm_lp_max: @p p -- power, @eps -- epsilon
+    /// #dnnl_reduction_norm_lp_add: @p p -- power, @eps -- epsilon
+    /// #dnnl_reduction_norm_lp_power_p_max: @p p -- power, @eps -- epsilon
+    /// #dnnl_reduction_norm_lp_power_p_add: @p p -- power, @eps -- epsilon
+    /// #dnnl_reduction_sum: @p p and @eps are ignored
+    /// #dnnl_reduction_mul: @p p and @eps are ignored
+    /// #dnnl_reduction_mean: @p p and @eps are ignored
+    float p, eps;
+} dnnl_reduction_desc_t;
+
+// dnnl.h
+dnnl_status_t DNNL_API dnnl_reduction_desc_init(
+        dnnl_reduction_desc_t *reduction_desc,
+        dnnl_alg_kind_t reduction_alg_kind, float p, float eps,
+        const dnnl_memory_desc_t *src_desc, const dnnl_memory_desc_t *dst_desc);
+```
+
+### Pros
+* Minimizes amount of post operations required to implement framework operations,
+ hence implementation is simpler.
+
+### Cons
+* New algorithms on framework side will require changes on oneDNN side.
+
+## Option 2: reduction and norms as combination of eltwise and reduction algorithms
+
+### API
+
+``` cpp
+// dnnl_types.h
+typedef enum {
+    ...
+    dnnl_eltwise_abs,
+    dnnl_eltwise_pow,
+    ...
+    dnnl_reduction_max,
+    dnnl_reduction_min,
+    dnnl_reduction_sum,
+    dnnl_reduction_mul,
+    ...
+} dnnl_alg_kind_t;
+
+typedef struct {
+    dnnl_primitive_kind_t primitive_kind;
+    // The kind of eltwise algorithm applied before reduction. This
+    // algorithm can be undefined.
+    dnnl_alg_kind_t eltwise_alg_kind;
+    // The kind of reduction algorithm.
+    dnnl_alg_kind_t reduction_alg_kind;
+    // Eltwise algorithm specific parameters.
+    float alpha, beta;
+    // Source memory descriptor.
+    dnnl_memory_desc_t src_desc;
+    // Destination memory descriptor.
+    dnnl_memory_desc_t dst_desc;
+} dnnl_reduction_desc_t;
+
+// dnnl.h
+dnnl_status_t DNNL_API dnnl_reduction_desc_init(
+        dnnl_reduction_desc_t *reduction_desc,
+        dnnl_alg_kind_t eltwise_alg_kind,
+        dnnl_alg_kind_t reduction_alg_kind,
```

Comment:
Two same `dnnl_alg_kind_t` types will certainly lead to erros of such API call. Only if reduction kind would be some what different type... But I would rather vote for an Option 1 since it's a bit more reliable for an end user.

Reply:
This is a minor thing on user side, because we have a documentation and API clearly states that algorithms are different (similar to multiple memory descriptors). The only problem i see is dispatching on the library side using algorithm.